### PR TITLE
[Fix] Execute and verify nested function calls.

### PR DESCRIPTION
**Pull request by d0cd**
This PR,
- fixes execution for nested function calls.
- fixes verification for nested function calls.
- adds tests using nested external calls and checks their execution order (the order of transitions in an `Execution`).

**To: Reviewer**
This PR needs feedback on how to appropriately organize the call graph reconstruction logic.

### What is the correct order?
Suppose that we have the following programs:
- `C`, with transition `c`.
- `B`, with transition `b`.
- `A`, which imports `B` and `C`, and contains transition `a` which invokes `B/b` and `C/c`.
Suppose that the call graph looks like:
```
A::a
  --> B::b
  --> C::c
```
Then the finalize blocks need to evaluated in the following order: `
```
B::b, C::c, A::a
```
In other words, **finalize blocks are evaluated in order of the transition that finished first**.

For a more complex example,
Suppose we have the following call graph, where each call is an external call.
```
A::a
--> B::b
       --> C::c
       --> D::d
--> E::e
       --> B::b
              --> C::c
              --> D::d"
       --> D::d
       --> C::c
```
The finalize blocks need to be evaluated in the following order: 
```
C::c, D::d, B::b, C::c, D::d, B::b, D::d, C::c, E::e, A::a
```

### CODE REVIEW

File: synthesizer/src/process/execute.rs [hunk 1]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+        }
+
+        impl<N: Network> TransitionMetadata<N> {
+            fn new(counter: &mut usize, pid: ProgramID<N>, fname: Identifier<N>, tid: Option<N::TransitionID>) -> Self {
+                let uid = *counter;
+                *counter += 1;
+                Self { uid, pid, fname, tid, children: None }
+            }
+
+            fn is_complete(&self) -> bool {
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs:262]:
```suggestion
            /// Returns 'true' if the subgraph starting from this transition has been fully-indexed.
            fn is_complete(&self) -> bool {
```

File: synthesizer/src/process/execute.rs [hunk 2]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+            Ok(())
+        };
+
+        // Initialize a call graph, which is a map of transition IDs to the transition IDs it calls.
+        let mut call_graph = HashMap::new();
+        // Initialize a mapping from UIDs to transition IDs.
+        let mut uid_to_tid = HashMap::new();
+
+        // Initialize a stack to track transition metadata, while traversing the call graph.
+        let mut traversal_stack: Vec<TransitionMetadata<N>> = Vec::new();
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs:289]: nit: rename to `metadata_stack` to improve readability.

File: synthesizer/src/process/execute.rs [hunk 3]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+    // In other words, we do not know which transitions were invoked by which transitions.
+    // Note that transition names are insufficient to reconstruct the call graph, since the same function can be invoked multiple times, in different ways.
+    //
+    // In order to reconstruct the call graph, we:
+    // - Iterate over the call structure in reverse post-order. The ordering is maintained by the `traversal_stack`.
+    // - Process each transition in the `Execution` in reverse, assigning its transition ID to the corresponding function call.
+    fn construct_call_graph(&self, execution: &Execution<N>) -> Result<HashMap<N::TransitionID, Vec<N::TransitionID>>> {
+        // Metadata for each transition the execution.
+        struct TransitionMetadata<N: Network> {
+            uid: usize,
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs:247]: nit: rename to `index` for consistency with rest of repo.

File: synthesizer/src/process/execute.rs [hunk 4]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+        // Iterate over each transition in reverse post-order, and populate the call graph.
+        for transition in execution.transitions().rev() {
+            // If the stack has complete metadata entries, then remove and add them to the call graph.
+            while let Some(metadata) = traversal_stack.last() {
+                if metadata.is_complete() {
+                    update_call_graph(traversal_stack.pop().unwrap(), &mut call_graph, &mut uid_to_tid)?;
+                } else {
+                    break;
+                }
+            }
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs]: This can be moved to the end of the `for` loop. Then the last `while` loop at the end of this function can be removed (and an `ensure!(traversal_stack.is_empty, "...")` can be added at the end of the function).

File: synthesizer/src/process/execute.rs [hunk 5]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+                    })
+                    .collect::<Vec<_>>();
+
+                // Add the children UIDs to the metadata.
+                // Note this unwrap is safe, for the same reason as above.
+                let top = traversal_stack.last_mut().unwrap();
+                let child_uids = children.iter().map(|child| child.uid).collect::<Vec<_>>();
+                match top.children {
+                    None => top.children = Some(child_uids),
+                    Some(ref mut children) => children.extend(child_uids),
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs]: To my knowledge, this `Some(..)` case is never hit. We should replace it with a `Some(..) => bail!("...")`.

**Review by howardwu — APPROVED** [review]
LGTM, pending minor safety changes

File: synthesizer/src/process/execute.rs [hunk 6]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+
+        // A helper function to update the call graph, given transition metadata.
+        let update_call_graph = |metadata: TransitionMetadata<N>,
+                                 call_graph: &mut HashMap<N::TransitionID, Vec<N::TransitionID>>,
+                                 uid_to_tid: &mut HashMap<usize, N::TransitionID>|
+         -> Result<()> {
+            // Check that the transition metadata is complete.
+            ensure!(metadata.is_complete(), "The transition metadata is incomplete");
+            // Update the UID to TID mapping.
+            uid_to_tid.insert(metadata.uid, metadata.tid.unwrap());
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs]: I would move this call to after `call_graph.insert()`. To my knowledge, the `call_graph.insert()`'s usage of `uid_to_tid[&uid]` on its `children` should fail if it ever needed this iteration's `uid_to_tid.insert()`.

File: synthesizer/src/process/execute.rs [hunk 7]
```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
...
+            // Update the call graph.
+            call_graph.insert(
+                metadata.tid.unwrap(),
+                metadata.children.unwrap().into_iter().map(|uid| uid_to_tid[&uid]).collect::<Vec<_>>(),
```
Comment (howardwu) [inline synthesizer/src/process/execute.rs]:
```suggestion
                                metadata
                    .children // Safe to unwrap, since the metadata is complete.
                    .unwrap()
                    .into_iter()
                    .map(|uid| uid_to_tid.get(&uid).copied().ok_or(anyhow!("Invalid traversal - missing 'tid' for uid '{uid}'")))
                    .collect::<Result<Vec<_>, _>>()?,
```

**Review by howardwu — APPROVED** [review]
LGTM
